pytensor/tensor/rewriting/math.py

```diff
@@ -167,6 +170,77 @@ def local_0_dot_x(fgraph, node):
             return [constant_zero]
 
 
+@register_stabilize
+@node_rewriter([Blockwise])
+def local_block_diag_dot_to_dot_block_diag(fgraph, node):
+    r"""
+    Perform the rewrite ``dot(block_diag(A, B), C) -> concat(dot(A, C), dot(B, C))``
+
+    BlockDiag results in the creation of a matrix of shape ``(n1 * n2, m1 * m2)``. Because dot has complexity
+    of approximately O(n^3), it's always better to perform two dot products on the smaller matrices, rather than
+    a single dot on the larger matrix.
+    """
+    if not isinstance(node.op.core_op, BlockDiagonal):
+        return
+
+    def check_for_block_diag(x):
+        return x.owner and (
+            isinstance(x.owner.op, BlockDiagonal)
+            or isinstance(x.owner.op, Blockwise)
+            and isinstance(x.owner.op.core_op, BlockDiagonal)
+        )
+
+    # Check that the BlockDiagonal is an input to a Dot node:
+    for client in get_clients_at_depth(fgraph, node, depth=1):
+        if not isinstance(client.op, Dot):
+            continue
+
+        op = client.op
+        x, y = client.inputs
+
+        if not (check_for_block_diag(x) or check_for_block_diag(y)):
+            continue
+
+        # Case 1: Only one input is BlockDiagonal. In this case, multiply all components of the block-diagonal with the
+        # non-block diagonal, and return a new block diagonal
+        if check_for_block_diag(x) and not check_for_block_diag(y):
+            components = x.owner.inputs
+            y_splits = split(
```

Comment:
Isn't this and the join along the 0th axis assuming a BlockwiseBlockDiagonal without batch dims?

Also not sure why you look for Dot but not Blockwise of _matrix_matrix_matmul. It doesn't always get rewritten as a Dot.

Reply:
I guess because you don't look for batch dot you can only have a BlockDiagonal without batch dims. That's fine but maybe a bit implicit. You can also wait for the useless BlockwiseBlockdiagonal to be rewritten as BlockDiagonal and only track that.

More importantly because you track a regular dot and not the matmul you may have a vector*matrix or matrix*vector product. Does the rewrite handle these correctly?

Reply:
If this is what's implied it's not on purpose. I'll modify it to account for blockwise dot.

There's no canonical `dot` form we rewrite to in an intermediate step to make reasoning about graphs easier? It seems nuts to have to to look for a bunch of different `_matrix_matrix_matmul` or `_matrix_vec_matmul` or `dot22` or whatever.

Reply:
I'm simplifying every blockwise as blockwise 2x2 dot (ie matmul) in https://github.com/pymc-devs/pytensor/pull/1471

The dot22 and dot22scalar are stuff from the blas pipeline and I've been hesitant to touch it. As first steps I would like to move them after specialize and to get rid of dot22scalar (should just be gemm). Those blas stuff should also work with blockwise but they currently don't.

Anyway if you target blockwise and core 2x2dot in this PR that should be the most robust going forward even if it misses some cases now. I suggest you explicitly exclude the vector matrix dots for now.

Reply:
skipping matrix-vector is a bit of a bummer

Reply:
you're like the teacher from that movie about jazz drumming u-u